internal/groups/groups.go

```diff
@@ -0,0 +1,111 @@
+package groups
+
+import (
+	"context"
+)
+
+type Member interface{}
+
+type Metadata map[string]interface{}
+
+type Group struct {
+	ID          string
+	OwnerID     string
+	ParentID    string
+	Name        string
+	Description string
+	Metadata    Metadata
+	Level       int
+	Path        string
+}
+
+type PageMetadata struct {
+	Total  uint64
+	Offset uint64
+	Limit  uint64
+	Name   string
+}
+
+type GroupPage struct {
+	PageMetadata
+	Groups []Group
+}
+
+type MemberPage struct {
+	PageMetadata
+	Members []Member
+}
+
+type Service interface {
+	// CreateGroup creates new  group.
+	CreateGroup(ctx context.Context, token string, group Group) (Group, error)
+
+	// UpdateGroup updates the group identified by the provided ID.
+	UpdateGroup(ctx context.Context, token string, group Group) (Group, error)
+
+	// ViewGroup retrieves data about the group identified by ID.
+	ViewGroup(ctx context.Context, token, id string) (Group, error)
+
+	// ListGroups retrieves groups.
+	ListGroups(ctx context.Context, token string, offset, limit uint64, m Metadata) (GroupPage, error)
+
+	// ListChildren retrieves groups that are children to group identified by parentID
+	ListChildren(ctx context.Context, token, parentID string, offset, limit uint64, m Metadata) (GroupPage, error)
+
+	// ListParents retrieves groups that are parent to group identified by childID.
+	ListParents(ctx context.Context, token, childID string, offset, limit uint64, m Metadata) (GroupPage, error)
```

Comment:
While I understand that using `parents` and `children` terminology is natural, I'm afraid it's also ambiguous because `parents` imply that one Group has multiple parents, and `children` implies only direct descendants.

Reply:
Also, did you consider adding a feature to search genealogy only up to some depth, something like: give me only my children? Or: give me my only my children and children of my children. Same for parents. There are many possibilities for searching group members, but let's not get into that in this PR, this is just a remark for future reference.

Reply:
https://github.com/mainflux/mainflux/pull/1259/files#r511910630, do Service methods return direct ascendents and descendants or the entire tree? This should be better explained in the documentation comments so that the one who implements knows the desirable behavior of the `Service` interface.